### DISCUSSION TITLE
Cannot import `node:crypto` in middleware.ts (placed in root dir as shown in the examples)

### DISCUSSION THREAD

**Discussion by 2jt** [vercel/next.js#47314] [Help] [answered]
### Summary

Hi. I'm able to use `import { createHmac } from 'node:crypto'` in `app/api/route.ts` file but importing it in /middleware.ts file I get this error:
```
node:crypto
Module build failed: UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs
```

### Additional information

```js
node version: v18.14.2
nextjs: "latest"
```

**Comment by rolanday** [comment]
Seeing the same when querying db using drizzle. Not a blocking issue for me (was just a convenient spot for me to exercise a test query) but begs the question re: limitation on middleware -- since this is server-side code path it's unclear why fails in middleware where same code block succeeds elsewhere.

**Reply by MuhammadBilal1234** [reply]
I am getting this error on querying and inserting using drizzle. what's the work around?

**Comment by csi-lk** [comment]
Same issue here

```
// ./middleware.ts

import * as crypto from 'node:crypto'
```

**Comment by rolanday** [comment]
For anybody tracking this, I believe the issue is that nextjs middleware does not support node and this may be by design, at least for now -- see https://nextjs.org/docs/messages/node-module-in-edge-runtime

**Reply by csi-lk** [reply]
This makes sense for when you're running on Vercel as it deploys the middleware to the 'edge' where the node apis may not exist, but for my usecase if running myself it should be allowed no?

**Reply by zkSoju** [reply]
Getting the same issue as well on newer NextJs versions but not on `13.1.6` so I'm curious if this is the root of the problem.

**Reply by jsadoski-rockhall** [reply]
I think the correct solution is to make a route (in the Node environment) and call it from your middleware.

**Reply by gcb** [reply]
and how vercel handled hmac on jwt all that time?

**Comment by MuhammadBilal1234** [comment]
any update on this issue? I am getting node:crypto error for drizzle as well

**Comment by nkg168** [comment]
The same error occurred when Next.js was at version 13.4.4, but upgrading to 13.4.7 fixed the issue.

**Reply by codecab** [reply]
Neither next.js version 13.4.7 nor 13.4.13 fixes this issue for me. I get the same error when trying to use drizzle from inside middleware.ts

**Reply by bigbizze** [reply]
I also get this error with 13.4.7 or 13.4.13 despite not using middleware / any edge functions

```
Module build failed: UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.

```

**Reply by nateschickler0** [reply]
failing for 13.4.13 here as well... unable to find solution

**Comment by StokleyO** [comment]
Any solution to this in place yet?

**Comment by rockyessel** [comment]
I had the same issue, just a few minutes ago. How i resolve d it was use "use server"

Drizzle ORM - What im using.

`src/lib/db/queries/blog.ts`
```js
'use server';

import db from '..';
import { blogs, roleBlog } from '../schema';
import { errorRes } from '@/lib/utils/constants';
import { BlogProps, ResObj, err } from '@/interface';
```
Error:
```sh
Module build failed: UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins (Unhandled scheme). Webpack supports "data:" and "file:" URIs by default. You may need an additional plugin to handle "node:" URIs.
```

Im assuning you are trying to make either a get or post http in a client components, if so, just put `use server`

**Reply by csi-lk** [reply]
Is this in the middleware specifically? Or just edge served queries?

**Reply by benjick** [reply]
Does not work in `middleware.ts`

**Comment by CasterCodes** [comment]
I had the same problem recently and adding  `"use server" ` on  top of the file where I run my queries helped

**Reply by csi-lk** [reply]
Is this in the `middleware.ts` specifically? Or just edge served queries?

**Comment by ShawnZeng** [comment]
"use server" does not work for me. In my case, I imported another package in  middleware.ts and in that package it is using node:crypto. Putting "use server" in middleware.ts and even the imported package does not work.

**Reply by nidrizi** [reply]
this happens to me too!  did you find any workaround or solution?

**Comment by coolbii** [comment]
same here
when adopt drizzle as orm
no matter what the nextjs version is,
it always happens to be this error

**Comment by floyd-may** [comment]
I'm encountering the same problem here, except I don't even have any middleware importing the server-only code. It's only reachable via API route and by instrumentation.

Edit: I just learned that the instrumentation hook is being evaluated for client-side use? I unwired the instrumentation hook and the error went away.

**Comment by kris-chateau** [comment]
Still not possible to use node:crypto with nextjs ?

**Comment by Jester0027** [comment]
What is the point of restricting nodejs modules inside the middleware ??
If you're already running it in nodejs, and deploying on a nodejs runtime too, why would that be a problem in the first place ?

**Reply by csi-lk** [reply]
As I posted above

> This makes sense for when you're running on Vercel as it deploys the middleware to the 'edge' where the node apis may not exist

**Reply by Jester0027** [reply]
I understand that it does some optimizations when deployed on Vercel, but there are still projects that are self-hosted and do run on NodeJS only. In that case it wouldn't make much sense to block native NodeJS modules.

**Reply by csi-lk** [reply]
Totally agree

> This makes sense for when you're running on Vercel as it deploys the middleware to the 'edge' where the node apis may not exist, but for my usecase if running myself it should be allowed no?

**Comment by wchorski** [comment]
think of the middleware as client side code. You'll have to do some /api/ route work around and fetch that endpoint in the middleware.

Here is my example middleware.ts with the idea of redirecting a route to a blog post from it's id to it's slug

btw I'm using Next 14 with app router

```ts
// cred - https://www.youtube.com/watch?v=xrvul-JrKFI
// other tips - https://borstch.com/blog/development/middleware-usage-in-nextjs-14
import { Post } from '@ks/types';
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'


export async function middleware(request: NextRequest) {

  //todo set user chosen theme here? from cookies

  if (request.nextUrl.pathname.startsWith('/blog/id')) {

    try {
      const url = new URL(request.url)
      const postId = url.pathname.split('/blog/id/')[1]

      const res = await fetch( url.origin + `/api/gql/noauth`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          query: `
            query Post($where: PostWhereUniqueInput!) {
              post(where: $where) {
                slug
              }
            }
          `,
          variables: {
            where: {
              id: postId
            }
          }
        }),
      })
      const {post} = await res.json() as {post: Post}
      if(!post.slug) return console.log('no post:slug found');
      
      return NextResponse.redirect(new URL(`/blog/${post?.slug}`, request.url))
      
    } catch (error) {
      return new Response('Error processing request', { status: 500 });
    }
  }
}
 
// See "Matching Paths" https://nextjs.org/docs/app/building-your-application/routing/middleware#matching-paths
export const config = {
  matcher:[ '/blog/id/:id*'],
}
```

**Comment by msobalvarro** [comment]
Any solution for this?

**Comment by jsadoski-rockhall** [comment]
Jumping in here to say that I think there are good design reasons to design middleware for the Node environment. I think the correct tack is to make a Route Handler with your Node-environment libraries like node:crypto, and call that route handler from your middleware.

**Comment by leerob** [comment, marked as answer]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
